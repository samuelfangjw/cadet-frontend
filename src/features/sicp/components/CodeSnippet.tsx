import { Card } from '@blueprintjs/core';
import * as React from 'react';
import { useState } from 'react';

import SicpWorkspace from './SicpWorkspace';

type CodeSnippetProps = OwnProps;
type OwnProps = {
  body: string;
  output: string;
};

const CodeSnippet: React.FC<CodeSnippetProps> = props => {
  const { body, output } = props;

  const [open, setOpen] = useState(false);

  const onClick = () => {
    setOpen(!open);
  };

  const WorkspaceProps = {
    editorValue: body,

    handleCloseEditor: onClick,
  }

  return (
    <div>
      {open ? (
        <SicpWorkspace {...WorkspaceProps}/>
      ) : (
        <>
          <Card className="sicp-code-snippet" onClick={onClick}>
            {body}
          </Card>
          {output ? <Card className="sicp-code-result">{output}</Card> : <></>}
        </>
      )}
    </div>
  );
};

export default CodeSnippet;
