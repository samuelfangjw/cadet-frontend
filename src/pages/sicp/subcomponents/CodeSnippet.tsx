import { HighlightRulesSelector, ModeSelector } from 'js-slang/dist/editors/ace/modes/source';
import * as React from 'react';
import AceEditor from 'react-ace';
import ControlBar from 'src/commons/controlBar/ControlBar';
import { ControlBarCloseButton } from 'src/commons/controlBar/ControlBarCloseButton';
import { ControlBarShowDependenciesButton } from 'src/commons/controlBar/ControlBarShowDependenciesButton';

import { CodeSnippetContext } from './SicpDisplay';
import SicpWorkspaceContainer from './SicpWorkspaceContainer';

type CodeSnippetProps = OwnProps;
type OwnProps = {
  body: string;
  output: string;
  id: string;
  initialEditorValueHash: string;
  initialPrependHash?: string | undefined;
  initialFullProgramHash?: string | undefined;
};

const CodeSnippet: React.FC<CodeSnippetProps> = props => {
  const { body, output, id } = props;
  const context = React.useContext(CodeSnippetContext);
  const [showPrepend, setShowPrepend] = React.useState(false);

  const handleShowDependencies = React.useCallback(() => {
    setShowPrepend(!showPrepend);
  }, [showPrepend]);

  const handleOpen = () => {
    context.setActive(id);
  };

  const handleClose = React.useCallback(() => {
    context.setActive('0');
  }, [context]);

  const WorkspaceProps = {
    initialEditorValueHash: showPrepend
      ? props.initialFullProgramHash
      : props.initialEditorValueHash,
    initialPrependHash: showPrepend ? undefined: props.initialPrependHash,
    isSicpEditor: true,

    handleCloseEditor: handleClose
  };

  HighlightRulesSelector(4);
  ModeSelector(4);

  const closeButton = React.useMemo(
    () => <ControlBarCloseButton key="close" handleClose={handleClose} />,
    [handleClose]
  );

  const showDependenciesButton = React.useMemo(
    () => (
      <ControlBarShowDependenciesButton
        key="dependencies"
        buttonText={showPrepend ? "Hide Dependencies" : "Show Dependencies"}
        handleShowDependencies={handleShowDependencies}
      />
    ),
    [handleShowDependencies, showPrepend]
  );

  const controlBarProps = {
    editorButtons: props.initialPrependHash ? [showDependenciesButton] : [],
    flowButtons: [],
    editingWorkspaceButtons: [closeButton]
  };

  const aceEditorProps = {
    className: 'react-ace',
    mode: 'source4defaultNONE',
    theme: 'source',
    fontSize: 20,
    highlightActiveLine: false,
    wrapEnabled: true,
    height: 'unset',
    width: '100%',
    showGutter: false,
    showPrintMargin: false,
    readOnly: true,
    maxLines: Infinity,
    value: body.replace(/\n$/, '') + ' ',
    setOptions: {
      fontFamily: "'Inconsolata', 'Consolas', monospace"
    }
  };

  return (
    <div className="sicp-code-snippet">
      {context.active === id ? (
        <>
          <ControlBar {...controlBarProps} />
          <SicpWorkspaceContainer {...WorkspaceProps} />
        </>
      ) : (
        <>
          <div className="code-body" onClick={handleOpen}>
            <AceEditor {...aceEditorProps}/>
          </div>
          {output ? (
            <pre className="code-result">
              <code>{output}</code>
            </pre>
          ) : (
            <></>
          )}
        </>
      )}
    </div>
  );
};

export default CodeSnippet;
