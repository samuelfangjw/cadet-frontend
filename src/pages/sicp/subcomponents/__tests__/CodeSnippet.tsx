import 'ace-builds';
import 'ace-builds/webpack-resolver';
import 'ace-builds/src-noconflict/ace';

import { shallow } from 'enzyme';
import lzString from 'lz-string';

import CodeSnippet from '../CodeSnippet';

describe('Sicp Code Snippet', () => {
  const body = '1+1;';
  const output = '2';
  const prependString = 'const a = 1;';
  const withoutPrepend = lzString.compressToEncodedURIComponent(body);
  const program = lzString.compressToEncodedURIComponent(prependString + '\n' + body);
  const prepend = lzString.compressToEncodedURIComponent(prependString);

  test('renders correctly', () => {
    const props = {
      body: body,
      output: output,
      id: 'id',
      initialEditorValueHash: withoutPrepend,
      initialPrependHash: prepend,
      initialFullProgramHash: program
    };

    const tree = shallow(<CodeSnippet {...props} />);
    expect(tree.debug()).toMatchSnapshot();
  });
});
