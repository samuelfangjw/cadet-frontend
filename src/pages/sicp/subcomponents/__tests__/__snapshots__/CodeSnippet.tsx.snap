// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sicp Code Snippet renders correctly 1`] = `
"<div className=\\"sicp-code-snippet\\">
  <div className=\\"code-body\\" onClick={[Function: handleOpen]}>
    <ReactAce className=\\"react-ace\\" mode=\\"source4defaultNONE\\" theme=\\"source\\" fontSize={20} highlightActiveLine={false} wrapEnabled={true} height=\\"unset\\" width=\\"100%\\" showGutter={false} showPrintMargin={false} readOnly={true} maxLines={Infinity} value=\\"1+1; \\" setOptions={{...}} name=\\"ace-editor\\" focus={false} enableSnippets={false} onChange={{...}} onPaste={{...}} onLoad={{...}} onScroll={{...}} minLines={{...}} tabSize={4} cursorStart={1} editorProps={{...}} style={{...}} scrollMargin={{...}} enableBasicAutocompletion={false} enableLiveAutocompletion={false} placeholder={{...}} navigateToFileEnd={true} />
  </div>
  <Component>
    2
  </Component>
</div>"
`;
